fem/pgridfunc.cpp

```diff
@@ -729,9 +737,17 @@ void ParGridFunction::ProjectBdrCoefficient(
 #endif
 }
 
+void ParGridFunction::ProjectBdrCoefficient(VectorCoefficient &vcoeff,
+                                            const Array<int> &attr)
+{
+   MFEM_VERIFY(VectorDim() == vcoeff.GetVDim(), "vcoeff vdim != VectorDim()");
+   ProjectBdrCoefficient(NULL, &vcoeff, attr);
+}
+
 void ParGridFunction::ProjectBdrCoefficientTangent(VectorCoefficient &vcoeff,
                                                    const Array<int> &bdr_attr)
 {
+   MFEM_VERIFY(VectorDim() == vcoeff.GetVDim(), "vcoeff vdim != VectorDim()");
```

Comment:
Here is the same problem as above with `ProjectBdrCoefficientNormal()`, it is just not caught by the test cases probably.

Reply:
This method is meant only for H(curl)-conforming spaces and their trace spaces, so the current check is correct.

Reply:
And the trace spaces require that "fix" again in `VectorDim()`. If there is some reason why something like `GetBdrVectorDim()` (or the optional parameter of `VectorDim()`) should not be implemented, I would prefer to at least remove the "fix" (which mixes elements of different dimensions) and manually check the space dimension similarly `ProjectBdrCoefficientNormal()`.

Reply:
I don't see a problem with supporting trace spaces in `FiniteElementSpace::GetVectorDim()`.

@pazner, @helloworld922, what do you think?

Reply:
The thing is it does not work and is not consistent 😬 . RTs in 2D have range dimension 2, while range dimension of 1D RT trace elements is 0. When you check correctly the range dimension for 1D face/boundary elements, both have range dimension 1. (The same for ND, just their trace dimension is 1).

Reply:
I don't understand your example for "does not work":

> while range dimension of 1D RT trace elements (picked by the "fix") is 0.

The trace space of RT is a scalar space (the trace here means the trace that makes sense -- the normal trace), so its dimension is 1. For example, for 2D `RT_FECollection`, the trace element is `L2_SegmentElement`. Since its `range_type` is `SCALAR` the implementation returns `GetVDim()` which is what we want.

Reply:
Just to clarify for Nedelec elements: for 2D `ND_FECollection`, the trace element is `ND_SegmentElement` which has `range_type` `VECTOR` and while its `RangeDim()` is 1, this is just the range in reference space. When mapped to physical space (with the H_CURL map, 1D -> 2D, or 1D -> 3D for a surface mesh) its dimension is `SpaceDimension()`. For 3D `ND_FECollection`, the trace is a 2D `ND_*Element` which has `RangeDim()` of 2, however, when mapped to physical space, its dimension is `SpaceDimension()`.